### ISSUE TITLE
Implement Test Driven Development

### ISSUE THREAD

**Issue by Iqrar99** [Iqrar99/WarnetBot#54]
This is gonna be a huge project because we need to make sure some of the logic is performing the way it should be.

![image](https://user-images.githubusercontent.com/20255031/220013513-ca4e2e21-8bd8-4e56-b47c-e7dc035435c8.png)

- [ ] Add GitHub action to test the code using `pytest`,
- [ ] Add Github action for coverage analysis.
- [ ] Add coverage test using [pytest-cov](https://pytest-cov.readthedocs.io/en/latest/),

**Comment by Iqrar99** [comment]
Reading resource:
- https://github.com/python-discord/bot/tree/main/tests

**Comment by rafiramadhana** [comment]
I think it is better to divide this issue into _sub-issues_

i.e.

- Setup CI/CD + module 1
- module 2
- module 3
- ...

So the PRs can be done per-modules

I am willing to contribute for initiating the unit tests

However, I need your help to prioritize which modules that should be test-covered first

**Comment by Iqrar99** [comment]
On the second thought, i've been researching about performing testing on discord bot till now. The blocker is it's still quite hard to figure out how to test the bot command manually since it involved discord API during the process. There are some limitations that the `discord.py` library can do.

But there is a library called [dpytest](https://github.com/CraftSpider/dpytest) that support this issue although it's still in development and pretty hard to use. Thus, we need to wait till everything is ok to continue the testing.

**Comment by rafiramadhana** [comment]
> Or perhaps we can move to the alternative solution. We create our own mocking object by using https://docs.python.org/3/library/unittest.mock.html (I prefer this way)

Agree, I think using a _generic_ mocking library like `unittest.mock` is a good idea

**Comment by rafiramadhana** [comment]
It looks like the core logic for this code base is located in `bot/cogs/...`. I think we can start from there for the TDD implementation.

**Comment by Iqrar99** [comment]
Closed for now
